File: bkbook/subbook.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2026 satorux
# Copyright (c) 1997-2006 Motoyuki Kasahara

"""サブブック——データファイルの先頭にある索引表。

``start`` / ``honmon`` の第 1 ページは、そのサブブック全体の目次になっている。
16 バイトの項目 1 つが 1 つの **索引** を指す。内容はページの範囲と、検索
できる索引ならば、検索語を引く前にどう正規化するかの規則である。項目の ID が
その範囲に何が入っているかを示す。本文そのもの、著作権表示、ある種類の
検索索引、あるいはフォント。

libeb の ``subbook.c`` にあたる。
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field

from .zio import EbError, Zio, find_directory, find_file

INDEX_TABLE_ENTRY_SIZE = 16

# 索引を引く前に、語をどう変形するか。
STYLE_CONVERT = 0
STYLE_ASIS = 1
STYLE_DELETE = 2

# フォントの種類。EB の索引 ID 0xf1..0xf8 が参照する順に並べてある。
FONT_16 = 0
FONT_24 = 1
FONT_30 = 2
FONT_48 = 3

FONT_HEIGHTS = {FONT_16: 16, FONT_24: 24, FONT_30: 30, FONT_48: 48}

#: 検索方式を表す索引 ID と、それを保持する属性名の対応。
SEARCH_INDEX_IDS = {
    0x00: "text",
    0x01: "menu",
    0x02: "copyright",
    0x10: "image_menu",
    0x70: "endword_kana",
    0x71: "endword_asis",
    0x72: "endword_alphabet",
    0x80: "keyword",
    0x81: "cross",
    0x90: "word_kana",
    0x91: "word_asis",
    0x92: "word_alphabet",
    0xD8: "sound",
}

#: multi 検索の表が、欄の語索引として使う索引 ID。ここに載る他のもの
#: （手元のディスクでは 0x01、0x05、0x0d）は本文のページである——
#: 入力できる値のメニューか、見出しを読み出す一覧のどちらか。
MULTI_INDEX_IDS = frozenset({0x71, 0x72, 0x91, 0x92, 0xA1, 0xA2, 0xB1, 0xB2})

#: そのうち、かなで鍵づけられているもの。0x70/0x90 の索引と同じように畳む。
MULTI_KANA_INDEX_IDS = frozenset({0xA1, 0xB1})

#: 内蔵フォントの開始ページを示す索引 ID。EB のみ。
FONT_INDEX_IDS = {
    0xF1: ("wide", FONT_16),
    0xF2: ("narrow", FONT_16),
    0xF3: ("wide", FONT_24),
    0xF4: ("narrow", FONT_24),
    0xF5: ("wide", FONT_30),
    0xF6: ("narrow", FONT_30),
    0xF7: ("wide", FONT_48),
    0xF8: ("narrow", FONT_48),
}


class _Unset:
    """遅延して求める値の印。``None`` を「答えなし」の意味に使えるようにする。"""


_UNSET = _Unset()


class SubBookError(EbError):
    """サブブックのデータファイルがないか、索引表が壊れている。"""


@dataclass
class Search:
    """索引 1 つ。どこにあるかと、そこ向けに語をどう正規化するか。"""

    index_id: int
    start_page: int
    end_page: int

    katakana: int = STYLE_ASIS
    lower: int = STYLE_CONVERT
    mark: int = STYLE_ASIS
    long_vowel: int = STYLE_ASIS
    double_consonant: int = STYLE_ASIS
    contracted_sound: int = STYLE_ASIS
    small_vowel: int = STYLE_ASIS
    voiced_consonant: int = STYLE_ASIS
    p_sound: int = STYLE_ASIS
    space: int = STYLE_DELETE

    @property
    def page_count(self) -> int:
        return self.end_page - self.start_page + 1


#: 欄に入力できる値を候補の木として並べた補助ページ。
MULTI_MENU_INDEX_ID = 0x01


@dataclass
class MultiEntry:
    """multi 検索の欄 1 つ。ラベルと、それに答える索引。"""

    label: str
    index: Search | None = None
    candidates: list[Search] = field(default_factory=list)

    @property
    def menu(self) -> Search | None:
        """入力できる値の木。本が用意している場合のみ。

        ことわざ の 事項 欄はまず 16 の見出しに分かれ、その各々がさらに
        10 ほどに分かれる。その木の葉が、この欄に実際に指定できる語である
        ——人生、一生。欄が挙げている他のページ（0x05、0x0d）はメニューでは
        なく、ヒットの見出しを読み出す一続きの並びである。
        """
        for search in self.candidates:
            if search.index_id == MULTI_MENU_INDEX_ID:
                return search
        return None


@dataclass
class MultiSearch:
    """語 1 つではなく、欄を 1 つずつ埋めて行う検索。

    語索引が「この見出し語に何が収められているか」に答えるのに対し、
    multi 検索は複数の問いを同時に立てる——ことわざ は 事項 と 作者 を、
    日本大百科 は 索引語 を求める——そしてそのすべてを満たす項目を返す。
    欄ごとに専用の索引があり、ヒットが指す項目の上で AND を取る。
    """

    search: Search
    entries: list[MultiEntry] = field(default_factory=list)

    @property
    def label(self) -> str:
        return " / ".join(entry.label for entry in self.entries if entry.label)


@dataclass
class Font:
    """内蔵フォント。ビットマップの開始位置と大きさ。"""

    font_code: int
    page: int = 0
    file_name: str = ""

    @property
    def height(self) -> int:
        return FONT_HEIGHTS[self.font_code]


class SubBook:
    """本の中の辞書 1 つ。"""

    def __init__(
        self,
        book,
        code: int,
        title: str,
        directory_name: str,
        index_page: int,
        narrow_font_names: list[str] | None = None,
        wide_font_names: list[str] | None = None,
    ):
        self.book = book
        self.code = code
        self.title = title
        self.directory_name = directory_name
        self.index_page = index_page

        self.searches: dict[str, Search] = {}
        self.multi_searches: list[Search] = []
        self.narrow_fonts: dict[int, Font] = {}
        self.wide_fonts: dict[int, Font] = {}
        self.search_title_page = 0

        self._narrow_font_names = narrow_font_names or []
        self._wide_font_names = wide_font_names or []
        self._zio: Zio | None = None
        self._font_zios: dict[str, Zio] = {}
        self._loaded = False
        self._multis: list[MultiSearch] | None = None
        self._stop_code: tuple[int, int] | None | _Unset = _UNSET

    # -- データファイル ----------------------------------------------------

    @property
    def directory_path(self) -> str:
        path = find_directory(self.book.path, self.directory_name)
        if path is None:
            raise SubBookError(
                f"{self.book.path}: no directory {self.directory_name!r} "
                f"for subbook {self.title!r}"
            )
        return path

    def _find_text_file(self) -> str:
        from .book import DISC_EB

        directory = self.directory_path
        if self.book.disc_code == DISC_EB:
            path = find_file(directory, "start")
            if path is None:
                raise SubBookError(f"{directory}: no start file")
            return path

        data_directory = find_directory(directory, "data") or directory
        for name in ("honmon", "honmon2"):
            path = find_file(data_directory, name)
            if path is not None:
                return path
        raise SubBookError(f"{data_directory}: no honmon file")

    @property
    def zio(self) -> Zio:
        """本文と索引のデータファイル。最初に使うときに開く。"""
        if self._zio is None:
            self._zio = Zio(self._find_text_file())
        return self._zio

    # -- 索引表 ------------------------------------------------------------

    def load(self) -> "SubBook":
        """索引表を解析する。何度呼んでもよい。必要なら自動的に呼ばれる。"""
        if self._loaded:
            return self

        page = self.zio.read_page(self.index_page)
        if len(page) < INDEX_TABLE_ENTRY_SIZE:
            raise SubBookError(f"{self.title}: truncated index table")

        count = page[1]
        if count == 0 or count >= len(page) // INDEX_TABLE_ENTRY_SIZE:
            raise SubBookError(f"{self.title}: bad index count {count}")

        global_availability = page[4]
        if global_availability > 0x02:
            global_availability = 0

        for i in range(count):
            offset = (i + 1) * INDEX_TABLE_ENTRY_SIZE
            self._parse_index_entry(
                page[offset : offset + INDEX_TABLE_ENTRY_SIZE], global_availability
            )

        self._attach_font_files()
        self._loaded = True
        return self

    def _search_from_entry(
        self, entry: bytes, global_availability: int
    ) -> Search | None:
        """16 バイトの索引表レコード 1 つから Search を組み立てる。

        multi 検索の表は、この同じ形のレコードを欄ごとに内部で繰り返している。
        そのレコードが **何であるか** の判定と分けてあるのはそのためである。
        """
        from .jacode import CHARCODE_ISO8859_1

        index_id = entry[0]
        start_page = int.from_bytes(entry[2:6], "big")
        block_count = int.from_bytes(entry[6:10], "big")
        if start_page == 0 or block_count == 0:
            return None

        search = Search(
            index_id=index_id,
            start_page=start_page,
            end_page=start_page + block_count - 1,
        )

        availability = entry[10]
        if (global_availability == 0x00 and availability == 0x02) or (
            global_availability == 0x02
        ):
            self._apply_style_flags(search, int.from_bytes(entry[11:14], "big"))
        elif index_id in (0x70, 0x90) or index_id in MULTI_KANA_INDEX_IDS:
            # かなの索引は、既定ですべてを畳む。
            search.katakana = STYLE_CONVERT
            search.lower = STYLE_CONVERT
            search.mark = STYLE_DELETE
            search.long_vowel = STYLE_CONVERT
            search.double_consonant = STYLE_CONVERT
            search.contracted_sound = STYLE_CONVERT
            search.small_vowel = STYLE_CONVERT
            search.voiced_consonant = STYLE_CONVERT
            search.p_sound = STYLE_CONVERT

        if self.book.character_code == CHARCODE_ISO8859_1 or index_id in (0x72, 0x92):
            search.space = STYLE_ASIS
        else:
            search.space = STYLE_DELETE
        return search

    def _parse_index_entry(self, entry: bytes, global_availability: int) -> None:
        from .book import DISC_EB

        search = self._search_from_entry(entry, global_availability)
        if search is None:
            return
        index_id = search.index_id
        start_page = search.start_page

        name = SEARCH_INDEX_IDS.get(index_id)
        if name is not None:
            self.searches[name] = search
        elif index_id == 0xFF:
            self.multi_searches.append(search)
        elif index_id == 0x16:
            if self.book.disc_code != DISC_EB:
                self.search_title_page = start_page
        elif index_id in FONT_INDEX_IDS and self.book.disc_code == DISC_EB:
            width, font_code = FONT_INDEX_IDS[index_id]
            fonts = self.wide_fonts if width == "wide" else self.narrow_fonts
            fonts[font_code] = Font(font_code=font_code, page=start_page)

    # -- multi 検索 --------------------------------------------------------

    @property
    def multis(self) -> list[MultiSearch]:
        """このサブブックの multi 検索。欄まで読み出したもの。

        索引表の 0xff レコードは索引を指しておらず、それ自身の小さな表を
        指している。中身はこうなっている: 2 バイトの欄数、14 バイトの空き、
        続いて欄ごとに、1 バイトの索引数、詰め物 1 バイト、30 バイトの
        ラベル、そしてその数だけの 16 バイト索引レコード——本体の索引表と
        同じ形のレコードである。そのうち 1 つがその欄の語索引で、残りは
        本文のページ（入力できる値のメニューと、ヒットの見出しを読み出す
        一覧）。
        """
        self.load()
        if self._multis is None:
            self._multis = [self._read_multi(s) for s in self.multi_searches]
        return self._multis

    def _read_multi(self, search: Search) -> MultiSearch:
        from .jacode import decode_jisx0208

        page = self.zio.read_page(search.start_page)
        count = int.from_bytes(page[0:2], "big")
        entries: list[MultiEntry] = []
        offset = 16
        for _field in range(count):
            if offset + 32 > len(page):
                raise SubBookError(f"{self.title}: multi search table overruns its page")
            index_count = page[offset]
            label_bytes = page[offset + 2 : offset + 32].split(b"\0")[0]
            entry = MultiEntry(label=decode_jisx0208(label_bytes))
            offset += 32
            for _index in range(index_count):
                record = page[offset : offset + INDEX_TABLE_ENTRY_SIZE]
                offset += INDEX_TABLE_ENTRY_SIZE
                index = self._search_from_entry(record, 0)
                if index is None:
                    continue
                if index.index_id in MULTI_INDEX_IDS:
                    entry.index = index
                else:
                    entry.candidates.append(index)
            entries.append(entry)
        return MultiSearch(search=search, entries=entries)

    @staticmethod
    def _apply_style_flags(search: Search, flags: int) -> None:
        """24 ビットの正規化フラグを Search に展開する。"""
        search.katakana = (flags & 0xC00000) >> 22
        search.lower = (flags & 0x300000) >> 20
        # この欄だけ反転している。0 が「記号を落とす」の意味。
        search.mark = STYLE_DELETE if (flags & 0x0C0000) >> 18 == 0 else STYLE_ASIS
        search.long_vowel = (flags & 0x030000) >> 16
        search.double_consonant = (flags & 0x00C000) >> 14
        search.contracted_sound = (flags & 0x003000) >> 12
        search.small_vowel = (flags & 0x000C00) >> 10
        search.voiced_consonant = (flags & 0x000300) >> 8
        search.p_sound = (flags & 0x0000C0) >> 6

    def _attach_font_files(self) -> None:
        """EPWING はフォントを、catalog に名前の書かれた別ファイルに持つ。

        そのファイルにはフォントが 1 つだけ入っているので、ヘッダは必ず
        第 1 ページにある。EB の本とは違う——あちらでは索引表が、共有の
        本文ファイル内で各フォントが始まるページを示している。
        """
        for fonts, names in (
            (self.narrow_fonts, self._narrow_font_names),
            (self.wide_fonts, self._wide_font_names),
        ):
            for font_code, name in enumerate(names):
                if not name:
                    continue
                font = fonts.setdefault(font_code, Font(font_code))
                font.file_name = name
                font.page = 1

    def font_zio(self, font: Font) -> Zio:
        """``font`` の字形が入っているデータファイル。"""
        if not font.file_name:
            return self.zio

        cached = self._font_zios.get(font.file_name)
        if cached is not None:
            return cached

        directory = find_directory(self.directory_path, "gaiji") or self.directory_path
        path = find_file(directory, font.file_name)
        if path is None:
            raise SubBookError(f"{directory}: no font file {font.file_name!r}")
        zio = Zio(path)
        self._font_zios[font.file_name] = zio
        return zio

    # -- 取り出し口 --------------------------------------------------------

    def search(self, name: str) -> Search | None:
        """名前で索引（``word_asis``、``text`` など）を返す。なければ None。"""
        self.load()
        return self.searches.get(name)

    # -- 項目がどこで終わるか ----------------------------------------------

    @property
    def stop_code(self) -> tuple[int, int] | None:
        """この辞書で項目を終わらせるエスケープシーケンス。

        最初に問われたときに本自身の索引から割り出す。1 秒ほどかかる。
        公開されている appendix のように答えが分かっているなら、代入して
        与えてもよい。
        """
        if isinstance(self._stop_code, _Unset):
            from . import stopcode

            self._stop_code = None  # 失敗を項目ごとに繰り返さないため
            try:
                self._stop_code = stopcode.infer(self)
            except EbError:
                pass
        return self._stop_code

    @stop_code.setter
    def stop_code(self, value: tuple[int, int] | None) -> None:
        self._stop_code = value

    @property
    def available_searches(self) -> list[str]:
        self.load()
        names = sorted(self.searches)
        if self.multi_searches:
            names.append(f"multi({len(self.multi_searches)})")
        return names

    def close(self) -> None:
        for zio in self._font_zios.values():
            zio.close()
        self._font_zios.clear()
        if self._zio is not None:
            self._zio.close()
            self._zio = None

    def __repr__(self) -> str:
        return f"<SubBook {self.code} {self.title!r} dir={self.directory_name!r}>"

File: web/bkbook/subbook.js
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2026 satorux
// Copyright (c) 1997-2006 Motoyuki Kasahara

/**
 * サブブック——データファイルの先頭にある索引表。Python 版 subbook.py の移植。
 */

import { CHARCODE_ISO8859_1, decodeJisx0208 } from "./jacode.js";
import { EbError, Zio, be, findDirectory, findFile } from "./zio.js";

const INDEX_TABLE_ENTRY_SIZE = 16;

// 索引を引く前に、語をどう変形するか。
export const STYLE_CONVERT = 0;
export const STYLE_ASIS = 1;
export const STYLE_DELETE = 2;

// フォントの種類。
export const FONT_16 = 0;
export const FONT_24 = 1;
export const FONT_30 = 2;
export const FONT_48 = 3;
export const FONT_HEIGHTS = { [FONT_16]: 16, [FONT_24]: 24, [FONT_30]: 30, [FONT_48]: 48 };

/** 検索方式を表す索引 ID と、それを保持する名前の対応。 */
export const SEARCH_INDEX_IDS = {
  0x00: "text",
  0x01: "menu",
  0x02: "copyright",
  0x10: "image_menu",
  0x70: "endword_kana",
  0x71: "endword_asis",
  0x72: "endword_alphabet",
  0x80: "keyword",
  0x81: "cross",
  0x90: "word_kana",
  0x91: "word_asis",
  0x92: "word_alphabet",
  0xd8: "sound",
};

/** multi 検索の表が、欄の語索引として使う索引 ID。 */
export const MULTI_INDEX_IDS = new Set([0x71, 0x72, 0x91, 0x92, 0xa1, 0xa2, 0xb1, 0xb2]);
/** そのうち、かなで鍵づけられているもの。 */
export const MULTI_KANA_INDEX_IDS = new Set([0xa1, 0xb1]);

/** 内蔵フォントの開始ページを示す索引 ID。EB のみ。 */
const FONT_INDEX_IDS = {
  0xf1: ["wide", FONT_16],
  0xf2: ["narrow", FONT_16],
  0xf3: ["wide", FONT_24],
  0xf4: ["narrow", FONT_24],
  0xf5: ["wide", FONT_30],
  0xf6: ["narrow", FONT_30],
  0xf7: ["wide", FONT_48],
  0xf8: ["narrow", FONT_48],
};

/** 欄に入力できる値を候補の木として並べた補助ページ。 */
const MULTI_MENU_INDEX_ID = 0x01;

/** サブブックのデータファイルがないか、索引表が壊れている。 */
export class SubBookError extends EbError {}

/** 索引 1 つ。どこにあるかと、そこ向けに語をどう正規化するか。 */
export class Search {
  constructor(indexId, startPage, endPage) {
    this.indexId = indexId;
    this.startPage = startPage;
    this.endPage = endPage;
    this.katakana = STYLE_ASIS;
    this.lower = STYLE_CONVERT;
    this.mark = STYLE_ASIS;
    this.longVowel = STYLE_ASIS;
    this.doubleConsonant = STYLE_ASIS;
    this.contractedSound = STYLE_ASIS;
    this.smallVowel = STYLE_ASIS;
    this.voicedConsonant = STYLE_ASIS;
    this.pSound = STYLE_ASIS;
    this.space = STYLE_DELETE;
  }

  get pageCount() {
    return this.endPage - this.startPage + 1;
  }
}

/** multi 検索の欄 1 つ。ラベルと、それに答える索引。 */
export class MultiEntry {
  constructor(label) {
    this.label = label;
    this.index = null;
    this.candidates = [];
  }

  get menu() {
    return this.candidates.find((search) => search.indexId === MULTI_MENU_INDEX_ID) || null;
  }
}

/** 語 1 つではなく、欄を 1 つずつ埋めて行う検索。 */
export class MultiSearch {
  constructor(search) {
    this.search = search;
    this.entries = [];
  }

  get label() {
    return this.entries.map((entry) => entry.label).filter(Boolean).join(" / ");
  }
}

/** 内蔵フォント。ビットマップの開始位置と大きさ。 */
export class Font {
  constructor(fontCode, page = 0, fileName = "") {
    this.fontCode = fontCode;
    this.page = page;
    this.fileName = fileName;
  }

  get height() {
    return FONT_HEIGHTS[this.fontCode];
  }
}

/** 本の中の辞書 1 つ。 */
export class SubBook {
  constructor(book, { code, title, directoryName, indexPage, narrowFontNames = [], wideFontNames = [] }) {
    this.book = book;
    this.fs = book.fs;
    this.code = code;
    this.title = title;
    this.directoryName = directoryName;
    this.indexPage = indexPage;

    this.searches = {};
    this.multiSearches = [];
    this.narrowFonts = {};
    this.wideFonts = {};
    this.searchTitlePage = 0;

    this._narrowFontNames = narrowFontNames;
    this._wideFontNames = wideFontNames;
    this._zio = null;
    this._fontZios = new Map();
    this._loading = null;
    this._multis = null;
    this.stopCode = undefined; // undefined = 未定、null = なし
    this._inferring = null;
  }

  /**
   * この辞書で項目を終わらせるエスケープシーケンス。[code, argument] か null。
   * 最初に問われたときに本自身の索引から割り出す（stopcode.js）。appendix の
   * ように答えが分かっているなら stopCode に代入しておけばよい。
   */
  async inferStopCode() {
    if (this.stopCode !== undefined) return this.stopCode;
    if (this._inferring === null) {
      this._inferring = (async () => {
        const { infer } = await import("./stopcode.js");
        try {
          this.stopCode = await infer(this);
        } catch (error) {
          if (!(error instanceof EbError)) throw error;
          this.stopCode = null; // 失敗を項目ごとに繰り返さないため
        }
        return this.stopCode;
      })();
    }
    return this._inferring;
  }

  // -- データファイル ------------------------------------------------------

  async directoryPath() {
    const path = await findDirectory(this.fs, this.book.path, this.directoryName);
    if (path === null) {
      throw new SubBookError(
        `${this.book.path}: no directory ${JSON.stringify(this.directoryName)} for subbook ${JSON.stringify(this.title)}`,
      );
    }
    return path;
  }

  async _findTextFile() {
    const directory = await this.directoryPath();
    if (this.book.discCode === "eb") {
      const path = await findFile(this.fs, directory, "start");
      if (path === null) throw new SubBookError(`${directory}: no start file`);
      return path;
    }
    const dataDirectory = (await findDirectory(this.fs, directory, "data")) || directory;
    for (const name of ["honmon", "honmon2"]) {
      const path = await findFile(this.fs, dataDirectory, name);
      if (path !== null) return path;
    }
    throw new SubBookError(`${dataDirectory}: no honmon file`);
  }

  /** 本文と索引のデータファイル。最初に使うときに開く。 */
  async zio() {
    if (this._zio === null) {
      this._zio = Zio.open(this.fs, await this._findTextFile()).catch((error) => {
        this._zio = null;
        throw error;
      });
    }
    return this._zio;
  }

  /** 1 起点のページ番号 page から count ページ読む。 */
  async readPage(page, count = 1) {
    return (await this.zio()).readPage(page, count);
  }

  // -- 索引表 ----------------------------------------------------------------

  /** 索引表を解析する。何度呼んでもよい。 */
  load() {
    if (this._loading === null) {
      this._loading = this._load().catch((error) => {
        this._loading = null;
        throw error;
      });
    }
    return this._loading;
  }

  async _load() {
    const page = await this.readPage(this.indexPage);
    if (page.length < INDEX_TABLE_ENTRY_SIZE) {
      throw new SubBookError(`${this.title}: truncated index table`);
    }
    const count = page[1];
    if (count === 0 || count >= Math.floor(page.length / INDEX_TABLE_ENTRY_SIZE)) {
      throw new SubBookError(`${this.title}: bad index count ${count}`);
    }
    let globalAvailability = page[4];
    if (globalAvailability > 0x02) globalAvailability = 0;

    for (let i = 0; i < count; i++) {
      const offset = (i + 1) * INDEX_TABLE_ENTRY_SIZE;
      this._parseIndexEntry(page.subarray(offset, offset + INDEX_TABLE_ENTRY_SIZE), globalAvailability);
    }
    this._attachFontFiles();
    return this;
  }

  /** 16 バイトの索引表レコード 1 つから Search を組み立てる。 */
  _searchFromEntry(entry, globalAvailability) {
    const indexId = entry[0];
    const startPage = be(entry, 2, 4);
    const blockCount = be(entry, 6, 4);
    if (startPage === 0 || blockCount === 0) return null;

    const search = new Search(indexId, startPage, startPage + blockCount - 1);
    const availability = entry[10];
    if ((globalAvailability === 0x00 && availability === 0x02) || globalAvailability === 0x02) {
      applyStyleFlags(search, be(entry, 11, 3));
    } else if (indexId === 0x70 || indexId === 0x90 || MULTI_KANA_INDEX_IDS.has(indexId)) {
      // かなの索引は、既定ですべてを畳む。
      search.katakana = STYLE_CONVERT;
      search.lower = STYLE_CONVERT;
      search.mark = STYLE_DELETE;
      search.longVowel = STYLE_CONVERT;
      search.doubleConsonant = STYLE_CONVERT;
      search.contractedSound = STYLE_CONVERT;
      search.smallVowel = STYLE_CONVERT;
      search.voicedConsonant = STYLE_CONVERT;
      search.pSound = STYLE_CONVERT;
    }
    if (this.book.characterCode === CHARCODE_ISO8859_1 || indexId === 0x72 || indexId === 0x92) {
      search.space = STYLE_ASIS;
    } else {
      search.space = STYLE_DELETE;
    }
    return search;
  }

  _parseIndexEntry(entry, globalAvailability) {
    const search = this._searchFromEntry(entry, globalAvailability);
    if (search === null) return;
    const indexId = search.indexId;
    const name = SEARCH_INDEX_IDS[indexId];
    if (name !== undefined) {
      this.searches[name] = search;
    } else if (indexId === 0xff) {
      this.multiSearches.push(search);
    } else if (indexId === 0x16) {
      if (this.book.discCode !== "eb") this.searchTitlePage = search.startPage;
    } else if (indexId in FONT_INDEX_IDS && this.book.discCode === "eb") {
      const [width, fontCode] = FONT_INDEX_IDS[indexId];
      const fonts = width === "wide" ? this.wideFonts : this.narrowFonts;
      fonts[fontCode] = new Font(fontCode, search.startPage);
    }
  }

  // -- multi 検索 ------------------------------------------------------------

  /** このサブブックの multi 検索。欄まで読み出したもの。 */
  async multis() {
    await this.load();
    if (this._multis === null) {
      const multis = [];
      for (const search of this.multiSearches) multis.push(await this._readMulti(search));
      this._multis = multis;
    }
    return this._multis;
  }

  async _readMulti(search) {
    const page = await this.readPage(search.startPage);
    const count = be(page, 0, 2);
    const multi = new MultiSearch(search);
    let offset = 16;
    for (let field = 0; field < count; field++) {
      if (offset + 32 > page.length) {
        throw new SubBookError(`${this.title}: multi search table overruns its page`);
      }
      const indexCount = page[offset];
      let labelBytes = page.subarray(offset + 2, offset + 32);
      const nul = labelBytes.indexOf(0);
      if (nul >= 0) labelBytes = labelBytes.subarray(0, nul);
      const entry = new MultiEntry(decodeJisx0208(labelBytes));
      offset += 32;
      for (let i = 0; i < indexCount; i++) {
        const record = page.subarray(offset, offset + INDEX_TABLE_ENTRY_SIZE);
        offset += INDEX_TABLE_ENTRY_SIZE;
        const index = this._searchFromEntry(record, 0);
        if (index === null) continue;
        if (MULTI_INDEX_IDS.has(index.indexId)) entry.index = index;
        else entry.candidates.push(index);
      }
      multi.entries.push(entry);
    }
    return multi;
  }

  _attachFontFiles() {
    for (const [fonts, names] of [
      [this.narrowFonts, this._narrowFontNames],
      [this.wideFonts, this._wideFontNames],
    ]) {
      names.forEach((name, fontCode) => {
        if (!name) return;
        if (!(fontCode in fonts)) fonts[fontCode] = new Font(fontCode);
        fonts[fontCode].fileName = name;
        fonts[fontCode].page = 1;
      });
    }
  }

  /** font の字形が入っているデータファイル。 */
  async fontZio(font) {
    if (!font.fileName) return this.zio();
    let cached = this._fontZios.get(font.fileName);
    if (cached === undefined) {
      const base = await this.directoryPath();
      const directory = (await findDirectory(this.fs, base, "gaiji")) || base;
      const path = await findFile(this.fs, directory, font.fileName);
      if (path === null) throw new SubBookError(`${directory}: no font file ${JSON.stringify(font.fileName)}`);
      cached = await Zio.open(this.fs, path);
      this._fontZios.set(font.fileName, cached);
    }
    return cached;
  }

  // -- 取り出し口 --------------------------------------------------------------

  /** 名前で索引（word_asis、text など）を返す。なければ null。 */
  async search(name) {
    await this.load();
    return this.searches[name] || null;
  }

  async availableSearches() {
    await this.load();
    const names = Object.keys(this.searches).sort();
    if (this.multiSearches.length) names.push(`multi(${this.multiSearches.length})`);
    return names;
  }

  async close() {
    for (const zio of this._fontZios.values()) await zio.close();
    this._fontZios.clear();
    if (this._zio !== null) {
      const zio = await this._zio.catch(() => null);
      if (zio) await zio.close();
      this._zio = null;
    }
  }
}

/** 24 ビットの正規化フラグを Search に展開する。 */
function applyStyleFlags(search, flags) {
  search.katakana = (flags & 0xc00000) >> 22;
  search.lower = (flags & 0x300000) >> 20;
  // この欄だけ反転している。0 が「記号を落とす」の意味。
  search.mark = (flags & 0x0c0000) >> 18 === 0 ? STYLE_DELETE : STYLE_ASIS;
  search.longVowel = (flags & 0x030000) >> 16;
  search.doubleConsonant = (flags & 0x00c000) >> 14;
  search.contractedSound = (flags & 0x003000) >> 12;
  search.smallVowel = (flags & 0x000c00) >> 10;
  search.voicedConsonant = (flags & 0x000300) >> 8;
  search.pSound = (flags & 0x0000c0) >> 6;
}
